### PR TITLE
[build-tools]: Use npm-check-updates in bump deps

### PR DESCRIPTION
**Pull request by tylerbutler**
This change refactors the `bump deps` command to use npm-check-updates for most of the logic to check npm for updates. This is already used by the `release` command, so this change unifies the code that checks for updates.

### CODE REVIEW

File: PACKAGES.md [hunk 1]
```diff
@@ -93,13 +93,13 @@ The dependencies between layers are enforced by the layer-check command._
 
 | Packages | Layer Dependencies |
```
Comment (tylerbutler) [inline PACKAGES.md:94]: Not sure why this change is here.

File: build-tools/packages/build-cli/src/base.ts [hunk 2]
```diff
@@ -157,6 +162,26 @@ export abstract class BaseCommand<T extends typeof BaseCommand.flags>
...
+            | undefined,
+    ): never;
+
+    public error(input: unknown, options?: unknown): void {
+        if (typeof input === "string") {
+            return super.error(chalk.red(input), options as any);
+        }
+
+        return super.error(input as Error, options as any);
+    }
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/base.ts:183]: Add docs - this is overriding the error methods in the oclif command class.

File: build-tools/packages/build-cli/src/lib/bump.ts [hunk 3]
```diff
@@ -15,6 +15,35 @@ import {
...
+ * A type guard used to determine if a string is a DependencyUpdateType.
+ *
+ * @internal
+ */
+export function isDependencyUpdateType(str: string | undefined): str is DependencyUpdateType {
+    if (str === undefined) {
+        return false;
+    }
+
+    if (["latest", "newest", "greatest", "minor", "patch"].includes(str)) {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/bump.ts:40]: Nit: Should we generate a static array of flags from the enum itself? Otherwise we'll have to update this any time we update the enum. I guess we would have to split the current type into 2 to do this reasonably though (1 for the pre-defined kinds, and 1 that includes the dynamic `@foo` kind). So feel free to disregard this :)

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 4]
```diff
@@ -15,14 +15,27 @@ import { VersionSpec } from "npm-check-updates/build/src/types/VersionSpec";
...
 import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
+import { DependencyUpdateType } from "./bump";
+
+/**
+ * An object that maps package names to version strings or range strings.
+ *
+ * @internal
+ */
+export interface PackageVersionMap {
+    [packageName: ReleasePackage]: string;
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:26]: Would it be worth adding a type-alias for the version / range string?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 5]
```diff
@@ -15,14 +15,27 @@ import { VersionSpec } from "npm-check-updates/build/src/types/VersionSpec";
...
+}
 
 /**
  * Checks the npm registry for updates for a release group's dependencies.
  *
  * @param context - The {@link Context}.
  * @param releaseGroup - The release group to check.
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:33]: We should probably document that `undefined` means update everything.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 6]
```diff
@@ -33,53 +46,86 @@ import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
 // eslint-disable-next-line max-params
 export async function npmCheckUpdates(
     context: Context,
-    releaseGroup: ReleaseGroup | ReleasePackage,
+    releaseGroup: ReleaseGroup | ReleasePackage | undefined,
     depsToUpdate: ReleasePackage[] | RegExp[],
-    bumpType: "patch" | "minor" | "current",
+    releaseGroupFilter: ReleaseGroup | undefined,
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:51]: Would it make sense for this to be an array of groups to filter? Or is there not a use-case for filtering out multiple?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 7]
```diff
@@ -33,53 +46,86 @@ import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
...
     const searchGlobs: string[] = [];
     let repoPath: string;
 
     log?.info(`Checking npm for package updates...`);
-    if (isReleaseGroup(releaseGroup)) {
+    if (releaseGroup === undefined) {
+        // Apply to all packages in repo
+        repoPath = context.repo.resolvedRoot;
+
+        for (const monorepo of context.repo.releaseGroups.values()) {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:77]: Nit: "currentReleaseGroup" (or something) instead of "monorepo"?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 8]
```diff
@@ -33,53 +46,86 @@ import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
...
+                ),
+                ".",
+            );
+        }
+
+        for (const pkg of context.independentPackages) {
+            searchGlobs.push(path.relative(context.repo.resolvedRoot, pkg.directory));
+        }
+
+        // searchGlobs.push(`./**`);
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:100]: Should this be removed?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 9]
```diff
@@ -33,53 +46,86 @@ import { isReleaseGroup, ReleaseGroup, ReleasePackage } from "../releaseGroups";
...
+    } else if (isReleaseGroup(releaseGroup)) {
         const monorepo = context.repo.releaseGroups.get(releaseGroup);
         if (monorepo === undefined) {
             throw new Error(`Can't find release group: ${releaseGroup}`);
         }
 
-        searchGlobs.push(...monorepo.workspaceGlobs);
         repoPath = monorepo.repoPath;
+        searchGlobs.push(...monorepo.workspaceGlobs, ".");
+        // searchGlobs.push(".");
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:109]: Should this be removed?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 10]
```diff
@@ -90,8 +136,31 @@ export async function npmCheckUpdates(
             throw new TypeError(`Expected an object: ${typeof result}`);
         }
 
-        for (const [pkgJsonPath, upgradedDeps] of Object.entries(result)) {
-            const jsonPath = path.join(repoPath, pkgJsonPath);
+        // npm-check-updates returns different data depending on how many packages were updated. This code detects the
+        // two main cases: a single package or multiple packages.
+        if (glob.endsWith("**")) {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:141]: Can't a glob ending with a single `*` also match multiple packages?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 11]
```diff
@@ -90,8 +136,31 @@ export async function npmCheckUpdates(
...
+            // if (numResults > 1) {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:142]: Should this be removed?

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 12]
```diff
@@ -90,8 +136,31 @@ export async function npmCheckUpdates(
...
+                // eslint-disable-next-line no-await-in-loop
+                const { name } = await readJsonAsync(jsonPath);
+                const pkg = context.fullPackageMap.get(name);
+                if (pkg === undefined) {
+                    log?.warning(`Package not found in context: ${name}`);
+                    continue;
+                }
+
+                for (const [dep, newRange] of Object.entries(upgradedDeps)) {
+                    upgradeLogLines.add(`    ${dep}: '${newRange}'`);
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/package.ts:154]: Nit: Wasn't there an indentation helper you added in another PR? Would it make sense to use that here?

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts [hunk 13]
```diff
@@ -184,6 +184,14 @@ export class GitRepo {
         return results.split("\n").filter(t => t !== undefined && t !== "" && t !== null);
     }
 
+    /**
+     * @returns An array containing all the modified files in the repo.
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:188]: Nit: Personally, I think it's potentially better to use the raw description field for comments that only specify what gets returned, rather than using an `@returns` block. That way intellisense / generated API docs, etc. see something useful as a summary. E.g. something as simple as:
```suggestion
     * Returns an array containing all the modified files in the repo.
```

**Review by Josmithr — APPROVED** [review]

File: build-tools/packages/build-cli/src/lib/bump.ts (hunk 3 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/bump.ts:40]: Yeah, if only union types were iterable (though how would you iterate the `@foo` ones?)! I was hoping to reuse a type from npm-check-updates but they don't export it, and it includes a function signature.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 7 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:77]: Went with `releaseGroupRoot`.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 6 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:51]: Hmmm, I can't think of a scenario for that at the moment, so I'll leave this for now. If a scenario comes up we can change it.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 5 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:33]: Good point.

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts (hunk 13 above)
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:199]: Good idea.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 12 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:154]: Yes! I completely missed this. Fixed.

File: build-tools/packages/build-cli/src/lib/package.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/package.ts:142]: Hmmm, good point. fluid-build only supports the `**` glob today, because that was all that was needed at the time. I'll change this.

File: build-tools/packages/version-tools/src/bumpTypes.ts [hunk 14]
```diff
@@ -5,6 +5,11 @@
 
 import type { SemVer } from "semver";
 
+/**
+ * A type alias for strings that represent package versions.
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/bumpTypes.ts:9]: Nit: any documentation regarding the form the contents of these strings are expected to take that we can link to from here?

**Review by Josmithr — APPROVED** [review]
